### PR TITLE
Add Generalize MoEGate module for WH

### PR DESCRIPTION
**Pull request by ronnie00TT**
### PR Category
Feature

### Summary
This PR builds a generalized MoEGate module for all the models listed in [#41132](https://github.com/tenstorrent/tt-metal/issues/41132), generalizing the existing optimized DeepSeek-V3 MoEGate to cover most of them.

#### Model side
1. Build a generalized, config-driven MoEGate module for all the [listed models](https://github.com/tenstorrent/tt-metal/tree/main/models/common/modules/moe/configs).
2. Support every model except qwen3.5-397B through the optimized MoEGate op.
3. Support a ttnn fallback path for qwen3.5-397B.
4. Update the per-model YAML config capturing each model's gate configuration.
5. Add tests for both the generalized MoEGate op and the full module.

This table lists how this op is generalized to each model and the perf time

| Model | Gate Operation Sequence | Supported | Notes | Wormhole Op (μs) | Wormhole Module (μs) |
|---|---|---|---|---|---|
| **DeepSeek V3** | linear → sigmoid → add bias → grouped TopK (8 groups → top8 experts) → gather → normalize & scale | ✅ Yes | — | 3.74 | 77.17 |
| **GPT OSS** | linear → add bias → TopK → softmax | ✅ Yes | Optimized op already exists (53.68 μs vs. 64.7 μs base) | 3.68 | 53.68 |
| **GLM-4.7** | linear → sigmoid → add bias → grouped TopK (1 group) → gather → normalize & scale | ✅ Yes | — | 4.20 | 66.48 |
| **GLM-5** | linear → sigmoid → add bias → grouped TopK (1 group) → gather → normalize & scale | ✅ Yes | — | 4.20 | 72.34 |
| **Kimi K2.5** | linear → sigmoid → add bias → grouped TopK (1 group, 384 experts) → gather → normalize & scale | ✅ Yes | — | 12.63 | 92.78 |
| **Qwen 3.5 (MoE, 397B)** | linear → softmax → TopK (512 experts, top10) → normalization | ❌ No | Softmax + Top10 + 512 experts not natively supported; falls back to standard ttnn ops | 294.21 | 397.44 |
| **Qwen 3.5 35B-A3B** | linear → softmax → TopK → normalization | ⚠️ Yes with workaround | Softmax is moved to the end of the op alongside normalization; mathematically equivalent but unverified with real inputs | 3.67 | 49.53 |
| **Qwen3 (MoE)** | linear → softmax → TopK → normalization | ⚠️ Yes with workaround | Softmax is moved to the end of the op alongside normalization; mathematically equivalent but unverified with real inputs | 3.69 | 60.51 |
| **DeepSeek OCR** | linear → softmax → TopK → normalize & scale | ⚠️ Yes with workaround | Softmax is moved to the end of the op alongside normalization; mathematically equivalent but unverified with real inputs | 3.71 | 44.25 |
| **Mistral Large 3** | linear → TopK → softmax | ✅ Yes | — | 3.69 | 76.89 |
| **Ling-MoE 1T** | linear → sigmoid → add bias → grouped TopK (8 groups → top8 experts) → gather → normalize & scale | ✅ Yes | — | 3.74 | 82.20 |
| **Qwen3-Omni 30B-A3B (thinker)** | linear → softmax → TopK → normalization | ⚠️ Yes with workaround | Softmax is moved to the end of the op alongside normalization; mathematically equivalent but unverified with real inputs | 3.68 | 49.38 |
| **Qwen3-Omni 30B-A3B (talker)** | linear → softmax → TopK (6 chosen) → normalization | ⚠️ Yes with workaround | Softmax is moved to the end of the op alongside normalization; mathematically equivalent but unverified with real inputs | 3.70 | 43.71 |
| **Gemma 4 26B-A4B** | RMSNorm → scale → linear → softmax → TopK → normalize & scale | ⚠️ Yes with workaround | Softmax moved to end; initial RMSNorm/scaling and per-expert post-scaling not natively supported by the module | 3.68 | 53.25<br>(Note: This time only covers from linear to normalize, excluding the final scale) |
| **DeepSeek V4 Pro** | linear → sqrtsoftplus → add bias → TopK → gather → normalize & scale | ⚠️ Yes with workaround | sqrtsoftplus is handled via a separate op within the module | 11.35 | 100.80 |
| **DeepSeek V4 Flash** | linear → sqrtsoftplus → add bias → TopK → gather → normalize & scale | ⚠️ Yes with workaround | sqrtsoftplus is handled via a separate op within the module | 3.61 | 70.58 |

### Test Results
Note that all the tests in `test_generalized_moe_gate.py` and `test_tt_moe_gate.py` passed on both BH and WH.

```
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-deepseek_ocr]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-deepseek_v3]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-deepseek_v4_flash]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-deepseek_v4_pro]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-gemma_4_26b]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-glm5]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-glm_47]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-gpt_oss]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-kimi_k25]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-ling_1t]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-mistral_large_3]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-qwen35_35b]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-qwen35_397b]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-qwen3_235b]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-qwen3_omni_talker]
PASSED models/common/tests/modules/moe/test_tt_moe_gate.py::test_tt_moe_gate[42-qwen3_omni_thinker]
```

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:ronnie/generalize_moe_gate_WH_module)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:ronnie/generalize_moe_gate_WH_module)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:ronnie/generalize_moe_gate_WH_module)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:ronnie/generalize_moe_gate_WH_module)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:ronnie/generalize_moe_gate_WH_module)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:ronnie/generalize_moe_gate_WH_module)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:ronnie/generalize_moe_gate_WH_module)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:ronnie/generalize_moe_gate_WH_module)

### CODE REVIEW

File: models/common/modules/moe/configs/deepseek_v4_flash.yaml [hunk 1]
```diff
@@ -8,6 +8,15 @@ num_routed_experts: 256
...
 has_bias: false
 
+# gate (routing front-end): n_group 1 = generalized op (ungrouped) / 8 = deepseek op (grouped top-8).
+# deepseek-v4: ungrouped (1 group) top-6 with the sqrtsoftplus score = sqrt(softplus(logit)) (applied
+# externally via ttnn.sqrt(ttnn.softplus(.)), no in-kernel op) + score_correction_bias for selection,
+# linear renormalization, scaled by 1.5. 256 experts → single 256-face (no combine). Same gate as
+# deepseek_v4_pro but 256 experts / scale 1.5 (pro is 384 / 2.5).
+n_group: 1
+score_func: sqrtsoftplus
+routed_scaling_factor: 1.5
```
[the remark this answers is not in the document]
Comment (ronnie00TT) [reply models/common/modules/moe/configs/deepseek_v4_flash.yaml:18]: Fixed

File: models/common/modules/moe/configs/deepseek_v4_pro.yaml [hunk 2]
```diff
@@ -8,6 +8,15 @@ num_routed_experts: 384
...
 has_bias: false
 
+# gate (routing front-end): n_group 1 = generalized op (ungrouped) / 8 = deepseek op (grouped top-8).
+# deepseek-v4: ungrouped top-6 with the sqrtsoftplus score = sqrt(softplus(logit)) (applied externally
+# via ttnn.sqrt(ttnn.softplus(.)), no in-kernel op) + score_correction_bias for selection, linear
+# renormalization, scaled by 2.5. 384 experts → the op's 512-combine path (pad 384→512, two 256-blocks →
+# global top-6), same combine path as kimi_k25.
+n_group: 1
+score_func: sqrtsoftplus
+routed_scaling_factor: 2.5
```
[the remark this answers is not in the document]
Comment (ronnie00TT) [reply models/common/modules/moe/configs/deepseek_v4_pro.yaml:18]: Fixed

File: models/common/modules/moe/configs/models_table.md [hunk 3]
```diff
@@ -2,8 +2,8 @@
 |---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|---|
 | **GPT-OSS 120B** | 2880 | 2880 | — | 128 | 0 | 4 | custom GELU-gated | softmax | simple | —/— | 1.0 | Yes | Yes | all MoE | 36 | No | GPT-OSS |
 | **DeepSeek V3** | 7168 | 2048 | 2048 | 256 | 1 | 8 | SiLU/SwiGLU | `sigmoid` | noaux_tc | 8/4 | 2.5 | No | correction | 3 | 61 | No | DS V3 |
-| **DS V4 Flash** | 4096 | 2048 | 2048 | 256 | 1 | 6 | SiLU/SwiGLU (swiglu_limit=10, routed only) | `sqrtsoftplus` (code) | noaux_tc; hash L0–2 (code) | —/— | 1.5 | No | correction (code) | 0 | 43 | No | DS V4 |
-| **DS V4 Pro** | 7168 | 3072 | 3072 | 384 | 1 | 6 | SiLU/SwiGLU (swiglu_limit=10, routed only) | `sqrtsoftplus` (code) | noaux_tc; hash L0–2 (code) | —/— | 2.5 | No | correction (code) | 0 | 61 | No | DS V4 |
+| **DS V4 Flash** | 4096 | 2048 | 2048 | 256 | 1 | 6 | SiLU/SwiGLU (swiglu_limit=10, routed only) | `sqrtsoftplus` (code) | noaux_tc; hash L0–2 (code) | —/— | 1.5 | No | No | 0 | 43 | No | DS V4 |
+| **DS V4 Pro** | 7168 | 3072 | 3072 | 384 | 1 | 6 | SiLU/SwiGLU (swiglu_limit=10, routed only) | `sqrtsoftplus` (code) | noaux_tc; hash L0–2 (code) | —/— | 2.5 | No | No | 0 | 61 | No | DS V4 |
```
[the remark this answers is not in the document]
Comment (ronnie00TT) [reply models/common/modules/moe/configs/models_table.md]: Fixed

File: models/common/modules/moe/tt_moe_gate.py [hunk 4]
```diff
@@ -0,0 +1,687 @@
...
+``_PAD_NEG`` experts that rank last); 257-512 uses the op's 2-block combine path (n_group=1 only).
+"""
+
+from __future__ import annotations
+
+import torch
+
+import ttnn
+from models.common.modules.moe.tt_moe_gate_config import TTMoEGateConfig
+from models.demos.deepseek_v3.tt.deepseek_moe_gate.op import DeepseekMoeGateOp
```
[the remark this answers is not in the document]
Comment (ronnie00TT) [reply models/common/modules/moe/tt_moe_gate.py]: Fixed

**Comment by yieldthought** [comment]
Codex review:

Findings:

1. [models/common/modules/moe/tt_moe_gate.py:570](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/modules/moe/tt_moe_gate.py#L570)
`TTMoEGate.forward` reshapes the routing tensors to `[1, 1, total_batch, k]`, but the common `TTMoEDecode` path requires the token, index, and score tensors to share the first three dimensions before dispatch. `all_to_all_dispatch_metadata` enforces that at [ttnn/cpp/ttnn/operations/experimental/ccl/all_to_all_dispatch_metadata/device/all_to_all_dispatch_metadata_device_operation.cpp:95](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/ttnn/cpp/ttnn/operations/experimental/ccl/all_to_all_dispatch_metadata/device/all_to_all_dispatch_metadata_device_operation.cpp#L95), and the existing common decode test fabricates routing as `[batch, 1, 1, k]` at [models/common/tests/modules/moe/test_tt_moe_decode.py:191](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/tests/modules/moe/test_tt_moe_decode.py#L191). With the normal decode input layout `[batch, 1, 1, hidden]`, wiring this new gate output directly into `TTMoEDecode.forward` will fail the shape check before dispatch. The new gate test avoids that by feeding a replicated `[1, 1, batch, hidden]` tensor and flattening the output, so it does not exercise the claimed router -> MoE composition. Suggestion: make `TTMoEGate` consume/return the common MoE layout `[batch, 1, 1, *]` or add an explicit adapter, then add a small gate -> decode integration test.

2. [models/common/tests/modules/moe/test_tt_moe_gate.py:44](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/tests/modules/moe/test_tt_moe_gate.py#L44)
The only module-level test for the new `TTMoEGate` is unconditionally skipped on Blackhole, while the module has arch/grid-dependent construction through `init_device_compute_kernel_config(self.mesh_device.arch())` at [models/common/modules/moe/tt_moe_gate.py:369](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/modules/moe/tt_moe_gate.py#L369) and a grid-derived matmul program config at [models/common/modules/moe/tt_moe_gate.py:405](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/modules/moe/tt_moe_gate.py#L405). That leaves the new wrapper, the 384/512 two-block paths, and the fallback path without Blackhole CI coverage even though the configs are being added as common model support. Suggestion: run a Blackhole-compatible subset of `test_tt_moe_gate` or keep the support explicitly WH/Galaxy-only until that follow-up lands.

**Comment by ronnie00TT** [comment]
> Codex review:
> 
> Findings:
> 
> 1. [models/common/modules/moe/tt_moe_gate.py:570](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/modules/moe/tt_moe_gate.py#L570)
>    `TTMoEGate.forward` reshapes the routing tensors to `[1, 1, total_batch, k]`, but the common `TTMoEDecode` path requires the token, index, and score tensors to share the first three dimensions before dispatch. `all_to_all_dispatch_metadata` enforces that at [ttnn/cpp/ttnn/operations/experimental/ccl/all_to_all_dispatch_metadata/device/all_to_all_dispatch_metadata_device_operation.cpp:95](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/ttnn/cpp/ttnn/operations/experimental/ccl/all_to_all_dispatch_metadata/device/all_to_all_dispatch_metadata_device_operation.cpp#L95), and the existing common decode test fabricates routing as `[batch, 1, 1, k]` at [models/common/tests/modules/moe/test_tt_moe_decode.py:191](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/tests/modules/moe/test_tt_moe_decode.py#L191). With the normal decode input layout `[batch, 1, 1, hidden]`, wiring this new gate output directly into `TTMoEDecode.forward` will fail the shape check before dispatch. The new gate test avoids that by feeding a replicated `[1, 1, batch, hidden]` tensor and flattening the output, so it does not exercise the claimed router -> MoE composition. Suggestion: make `TTMoEGate` consume/return the common MoE layout `[batch, 1, 1, *]` or add an explicit adapter, then add a small gate -> decode integration test.
> 2. [models/common/tests/modules/moe/test_tt_moe_gate.py:44](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/tests/modules/moe/test_tt_moe_gate.py#L44)
>    The only module-level test for the new `TTMoEGate` is unconditionally skipped on Blackhole, while the module has arch/grid-dependent construction through `init_device_compute_kernel_config(self.mesh_device.arch())` at [models/common/modules/moe/tt_moe_gate.py:369](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/modules/moe/tt_moe_gate.py#L369) and a grid-derived matmul program config at [models/common/modules/moe/tt_moe_gate.py:405](https://github.com/tenstorrent/tt-metal/blob/d2cda76df2a604163d6bd3bf72d1854cc1df847e/models/common/modules/moe/tt_moe_gate.py#L405). That leaves the new wrapper, the 384/512 two-block paths, and the fallback path without Blackhole CI coverage even though the configs are being added as common model support. Suggestion: run a Blackhole-compatible subset of `test_tt_moe_gate` or keep the support explicitly WH/Galaxy-only until that follow-up lands.

@yieldthought Thanks for your reviews, both have been fixed now.

**Review by yieldthought — APPROVED** [review]
